installdir/hlload.1.in
.TH hlload 1 "2018-04-04"
.\" NAME should be all caps, SECTION should be 1-8, maybe w/ subsection
.\" other parms are allowed: see man(7), man(1)
.SH NAME
hlload \- reads in an ASCII-text version of the document database
.SH SYNOPSIS
.B hlload
.I "[options]"
.SH "DESCRIPTION"
hlload  reads in an ASCII-text version of the document database in the same form as
the  -t  option  of  hldig  and  hldump. Note that this will overwrite data in your
databases, so this should be used with great care.
.SH OPTIONS
.TP
.B \-a
Use alternate work files. Tells hlload to append 
.I .work
to database files, allowing it to
operate on a second set of databases.
.TP
.B \-c \fIconfigfile\fR
Use the specified
.I configfile
instead of the default.
.TP
.B \-i
Initial.  Do  not use any old databases. This is accomplished by first erasing
the databases.
.TP
.B \-v
Verbose mode. This doesn't have much effect.
.SH File Formats
.TP
.B Document Database
Each line in the file starts with the document id 
followed by a list of
.I fieldname
:
.I value
separated by tabs. The fields always appear in the
order listed below:
.TP
.B u
URL
.TP
.B t
Title
.TP
.B a
State (0 = normal, 1 = not found, 2 = not indexed, 3 = obsolete)
.TP
.B m
Last modification time as reported by the server
.TP
.B s
Size in bytes
.TP
.B H
Excerpt
.TP
.B h
Meta description
.TP
.B l
Time of last retrieval
.TP
.B L
Count of the links in the document (outgoing links)
.TP
.B b
Count of the links to the document (incoming links or backlinks)
.TP
.B c
HopCount of this document
.TP
.B g
Signature of the document used for duplicate-detection
.TP
.B e
E-mail address to use for a notification message from hlnotify
.TP
.B n
Date to send out a notification e-mail message
.TP
.B S
Subject for a notification e-mail message
.TP
.B d
The text of links pointing to this document. (e.g. <a href="docURL">description</a>)
.TP
.B A
Anchors in the document (i.e. <A NAME=...)
.TP
.B Word Database
While hldump and hlload don't deal with the word database
directly, it's worth mentioning it here because you need to
deal with it when copying the ASCII databases from one system
to another. The initial word database produced by hldig is
already in ASCII format, and a binary version of it is
produced by hlmerge, for use by hlsearch. So, when you copy
over the ASCII version of the document database produced by
hldump, you need to copy over the wordlist as well, then
run hlload to make the binary document database on the target
system, followed by running hlmerge to make the word index.
.TP
Each line in the word list file starts with the word
followed by a list of
.I fieldname
:
.I value
separated by tabs. The fields always appear in the
order listed below, with the last two being optional:
.TP
.B i
Document ID
.TP
.B l
Location of word in document (1 to 1000)
.TP
.B w
Weighl of word based on scoring factors
.TP
.B c
Count of word's appearances in document, if more than 1
.TP
.B a
Anchor number if word occurred after a named anchor
.SH "FILES"
.TP
.B @DEFAULT_CONFIG_FILE@
The default configuration file.
.TP
.B @DATABASE_DIR@/db.docs
The default ASCII document database file.
.TP
.B @DATABASE_DIR@/db.wordlist
The default ASCII word database file.
.SH "SEE ALSO"
Please refer to the HTML pages (in the hldig-doc package)
.B /usr/share/doc/hldig-doc/html/index.html
and the manual pages
.I hldig(1)
,
.I hlmerge(1)
and
.I hldump(1)
for a detailed description of hl://Dig and its commands.
.SH AUTHOR
This manual page was written by Stijn de Bekker, based on the HTML
documentation of hl://Dig.


ChangeLog
2018-04-04  Johnny A. Solbu <johnny@solbu.net>

	* Remove broken build test link
	* rename ht -> hl in topdir files
	* Complete rename in installdir
	* Autoreconf
	* Apparently several subscripts still use the VERSION file. Should
	fix that sometime
	* Rename mans in configure and make scripts.
	* Rename man files and their contents

2018-04-02  Johnny A. Solbu <johnny@solbu.net>

	* VERSION file is obsolete now that version is in configure.ac
	[skip ci]
	* Update pot file: copyright are assigned to «the hl://Dig
	group» [skip ci]
	* commit fb7ea686ffe3fd4aac7e823f1855fb34a3f68dd3 Author: Johnny
	A. Solbu <johnny@solbu.net> Date:   Mon Apr 2 13:31:43 2018 +0200

2018-03-27  andy5995 <andy400-dev@yahoo.com>

	* dbbench.cc:fix conversion warning This should fix the warning when `make dist` is run: dbbench.cc:269:20: warning: ISO C++ forbids converting a string
	constant to ‘char*’ [-Wpedantic]    char
	*progname = "dbbench problem...";

2018-03-28  Johnny A. Solbu <johnny@solbu.net>

	* bump version: 0.2.0
	* Add bug address to configure
	* Put version in the configure.ac file, where it belongs
	* Update build tests location
	* Rename url in the last files

2018-03-27  Johnny A. Solbu <johnny@solbu.net>

	* Rename urls in more files
	* Change MSGID_BUGS_ADDRESS, point to the issue tracker
	* Update ChangeLog.md with the new URLs
	* New project management: Rename URLs
	* Don't track compiled translations in revision controll
	* Remove unneeded preceding slash

2018-03-27  andy5995 <andy400-dev@yahoo.com>

	* change htdig.gif to logo.jpg (#56) After changes by @jamessouth, I found a few other areas I forgot to
	mention to him.

2018-03-15  jamessouth <jamesdanielsouth@gmail.com>

	* issue 56 new logo and modified html files

2018-03-15  Andy Alt <andy5995@users.noreply.github.com>

	* Changlog.md:add note about default install paths [skip ci]
	* Makefile.am:remove CVS dist-hook [skip ci]
	* CONTRIBUTING.md:add note about "skip ci" [skip ci]
	* TESTING.md:updated [skip ci] This reflects changes to the new install locations, which was a
	recent patch submitted by @Solbu
	* push regenerated build scripts and makefiles Regenerated due to previous commit
	* configure.ac:change path of www images [skip ci] When installdir/Makefile is run, it uses the value here to prefix
	urls to images in the installdir/html templates with /www/ To be accessible, a symlink can be put in the web root that points
	to that directory.

2018-03-15  Johnny A. Solbu <github@solbu.net>

	* Install things in the proper places (#90)
	  Install things in PREFIX/share/hldir, rather than /var/www

2018-03-12  Andy Alt <andy5995@users.noreply.github.com>

	* Merge pull request #101 from solbu/installer-fixes Installer fixes (Issue #99)
	* republish site from changes in bcdf311 This republishes the site using changes @jamessouth made in
	bcdf3110c78a308404431f3f3a24d33503d283a5 (#81)

2018-03-12  james south <jamessouth@users.noreply.github.com>

	* make sure home page validates (#100) (fixes #81) [skip ci]

2018-03-12  Johnny A. Solbu <johnny@solbu.net>

	* Add ABOUT-NLS to distributed files, indent traling backslash
	* Add missing gettext.h header

2018-03-11  andy5995 <andy400-dev@yahoo.com>

	* ABOUT-NLS:add to distribution [skip ci]
	* remove '#' (forgot to do this in a previous commit) [skip ci]
	* .travis.yml:drop '#' from hldig Freenode channel [skip ci]
	* simplecgen:refactor trim() function
	* list #hldig for the chat room, not ##hldig [skip ci]
	* .travis.yml: change failure and error notifications Only notify on_failure if there's a change Always notify on error
	* .travis.yml:change irc notification options
	* .travis.yml:change osx and Xcode version (#98)
	* .travis.yml:change osx and Xcode version
	* .travis.yml:add matrix closes #76

2018-03-11  Johnny A. Solbu <github@solbu.net>

	* Fix test dir (issue #91) (#97) / Thanks @solbu
	* Fix «make dist» Some tools where renamed at some point but the test commands using
	them where not updated
	* Update test/test_function.in with the renamed filenames

2018-03-09  Andy Alt <andy5995@users.noreply.github.com>

	* Merge pull request #92 from solbu/doc-changes [skip ci] Simplify translation instructions

2018-03-07  andy5995 <andy400-dev@yahoo.com>

	* translation:add ru to LINGUAS, generate gmo file (closes #68)
	* travis.yml:add osx (#76)
	* ne.po:remove empty file [skip ci]

2018-03-07  Svitlana Galianova <svitlana.galianova@gmail.com>

	* added russian translation

2018-03-07  Johnny A. Solbu <github@solbu.net>

	* Uninstall fix (#84) (closes #48)
	* Downgrade required gettext version, to match my system
	* Add Solbu to authors list

2018-03-06  andy5995 <andy400-dev@yahoo.com>

	* ChangeLog.md:note about html docs not being installed anymore

2018-03-06  Johnny A. Solbu <johnny@solbu.net>

	* Run autoreconf
	* Don't install documentaion in docdir RPM packagers prefer to control that dir using a «%doc» macro
	* Ignore .patch files To avoid committing patches by mistake, when one is also working on
	packaging

2018-01-21  andy5995 <andy400-dev@yahoo.com>

	* fix formatting [skip ci]
	* remove Slack, use IRC found some leftovers to remove (see previous commit)
	* remove Slack workspace from contact info [skip ci] IRC should be sufficient for this project.  http://webchat.freenode.net?channels=%23%23hldig&uio=d4

2018-01-15  andy5995 <andy400-dev@yahoo.com>

	* docs/README: explain how to view the html files

2018-01-11  Andy Alt <andy5995@users.noreply.github.com>

	* [#24]installdirs:update search result html, add css (#82)
	* [#24]installdirs:update search result html, add css This will copy the same style.css that the web site uses. That seems
	easier than maintaining 2 separate css files.  I added a script that runs tidy-html5, but the user must install the
	program <http://www.htacg.org/> to use it. 
	* hlsearch results html pages updated [#24] See notes in the

hltools/hlload.cc
//
// htload.cc
//
// htload: A utility to read ASCII text versions of the document
// and/or word databases. These can be used by external programs,
// edited, or used as a platform and version-independent form of the DB.
//
// Part of the ht://Dig package   <http://www.htdig.org/>
// Copyright (c) 1999-2004 The ht://Dig Group
// For copyright details, see the file COPYING in your distribution
// or the GNU Library General Public License (LGPL) version 2 or later
// <http://www.gnu.org/copyleft/lgpl.html>
//
// $Id: htload.cc,v 1.6 2004/05/28 13:15:25 lha Exp $
//

#ifdef HAVE_CONFIG_H
#include "htconfig.h"
#endif /* HAVE_CONFIG_H */

#include "WordContext.h"
#include "HtURLCodec.h"
#include "HtWordList.h"
#include "HtConfiguration.h"
#include "DocumentDB.h"
#include "defaults.h"
#include "messages.h"

#include <errno.h>

#ifndef _MSC_VER                /* _WIN32 */
#include <unistd.h>
#endif

// If we have this, we probably want it.
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#elif HAVE_GETOPT_LOCAL
#include <getopt_local.h>
#endif

int verbose = 0;

void usage ();
void reportError (char *msg);

//*****************************************************************************
// int main(int ac, char **av)
//
int
main (int ac, char **av)
{
  int do_words = 1;
  int do_docs = 1;
  int alt_work_area = 0;
  String configfile = DEFAULT_CONFIG_FILE;
  int c;
  extern char *optarg;

  while ((c = getopt (ac, av, "vdwc:a")) != -1)
  {
    switch (c)
    {
    case 'c':
      configfile = optarg;
      break;
    case 'v':
      verbose++;
      break;
    case 'a':
      alt_work_area++;
      break;
    case 'w':
      do_words = 0;
      break;
    case 'd':
      do_docs = 0;
      break;
    case '?':
      usage ();
      break;
    }
  }

  HtConfiguration *config = HtConfiguration::config ();
  config->Defaults (&defaults[0]);

  if (access ((char *) configfile, R_OK) < 0)
  {
    reportError (form (_("Unable to find configuration file '%s'"),
                       configfile.get ()));
  }

  config->Read (configfile);

  //
  // Check url_part_aliases and common_url_parts for
  // errors.
  String url_part_errors = HtURLCodec::instance ()->ErrMsg ();

  if (url_part_errors.length () != 0)
    reportError (form (_("Invalid url_part_aliases or common_url_parts: %s"),
                       url_part_errors.get ()));


  // We may need these through the methods we call
  if (alt_work_area != 0)
  {
    String configValue;

    configValue = config->Find ("word_db");
    if (configValue.length () != 0)
    {
      configValue << ".work";
      config->Add ("word_db", configValue);
    }

    configValue = config->Find ("doc_db");
    if (configValue.length () != 0)
    {
      configValue << ".work";
      config->Add ("doc_db", configValue);
    }

    configValue = config->Find ("doc_index");
    if (configValue.length () != 0)
    {
      configValue << ".work";
      config->Add ("doc_index", configValue);
    }

    configValue = config->Find ("doc_excerpt");
    if (configValue.length () != 0)
    {
      configValue << ".work";
      config->Add ("doc_excerpt", configValue);
    }
  }

  if (do_docs)
  {
    const String doc_list = config->Find ("doc_list");
    DocumentDB docs;
    if (docs.Open (config->Find ("doc_db"), config->Find ("doc_index"),
                   config->Find ("doc_excerpt")) == OK)
    {
      docs.LoadDB (doc_list, verbose);
      docs.Close ();
    }
  }
  if (do_words)
  {

    // Initialize htword
    WordContext::Initialize (*config);

    const String word_dump = config->Find ("word_dump");
    HtWordList words (*config);
    if (words.Open (config->Find ("word_db"), O_RDWR) == OK)
    {
      words.Load (word_dump);
      words.Close ();
    }
  }

  return 0;
}


//*****************************************************************************
// void usage()
//   Display program usage information
//
void
usage ()
{
  Usage help;
  cout << _("usage:");
  cout << " hlload [-v][-d][-w][-a][-c configfile]\n";
  printf (_("This program is part of hl://Dig %s\n\n"), VERSION);
  cout << _("Options:\n");

  help.verbose ();

  cout << "\t-d\tDo NOT load the document database.\n\n";
  cout << "\t-w\tDo NOT load the word database.\n\n";

  help.alternate_common ();
  cout << "\t\tTells hlload to append .work to the database files \n";
  cout << "\t\tallowing it to operate on a second set of databases.\n";

  help.config ();
  exit (0);
}


//*****************************************************************************
// Report an error and die
//
void
reportError (char *msg)
{
  cout << "hlload: " << msg << "\n\n";
  exit (1);
}
